libclc/libspirv/lib/ptx-nvidiacl/group/collectives_helpers.ll

```diff
@@ -1,97 +1,73 @@
 ; 32 storage locations is sufficient for all current-generation NVIDIA GPUs
-; 128 bits per warp is sufficient for all fundamental data types and complex
-; Reducing storage for small data types or increasing it for user-defined types
-; will likely require an additional pass to track group algorithm usage
-@__clc__group_scratch = internal addrspace(3) global [128 x i64] undef, align 1
+@__clc__group_scratch_i1 = internal addrspace(3) global [32 x i1] poison, align 1
+@__clc__group_scratch_i8 = internal addrspace(3) global [32 x i8] poison, align 1
+@__clc__group_scratch_i16 = internal addrspace(3) global [32 x i16] poison, align 2
+@__clc__group_scratch_i32 = internal addrspace(3) global [32 x i32] poison, align 4
+@__clc__group_scratch_i64 = internal addrspace(3) global [32 x i64] poison, align 8
+@__clc__group_scratch_i128 = internal addrspace(3) global [32 x i128] poison, align 8
 
-define i8 addrspace(3)* @__clc__get_group_scratch_bool() nounwind alwaysinline {
+define ptr addrspace(3) @__clc__get_group_scratch_bool() nounwind alwaysinline {
 entry:
-  %ptr = getelementptr inbounds [128 x i64], [128 x i64] addrspace(3)* @__clc__group_scratch, i64 0, i64 0
-  %cast = bitcast i64 addrspace(3)* %ptr to i8 addrspace(3)*
-  ret i8 addrspace(3)* %cast
+  %0 = getelementptr inbounds [32 x i1], ptr addrspace(3) @__clc__group_scratch_i1, i64 0, i64 0
```

Comment:
This is equivalent to simply `ret ptr addrspace(3) @__clc_group_scratch_i1`. I don't think we need these `getelementptr`s at all.

Reply:
That's why I'm not sure we even really need the `_<TYPE>` functions anymore. We could probably just have one unified scratch helper.

Reply:
> This is equivalent to simply `ret ptr addrspace(3) @__clc_group_scratch_i1`. I don't think we need these `getelementptr`s at all.

done, thanks

Reply:
> That's why I'm not sure we even really need the `_<TYPE>` functions anymore. We could probably just have one unified scratch helper.

Using multiple global variable with different sizes can solve the overestimation of local variable size issue, as mentioned in deleted comment by this PR. For instance, if a test is using char type, there is no need to use a local variable of size 32 x i128.